src/CFamily/CMake/CompilationConfigProvider.cs

```diff
@@ -109,5 +120,97 @@ public CompilationDatabaseEntry GetConfig(string filePath)
                 return null;
             }
         }
+
+        private bool IsHeaderFile(string filePath) => 
+            Path.GetExtension(filePath).Equals(".h", StringComparison.OrdinalIgnoreCase);
+
+        private CompilationDatabaseEntry LocateExactCodeEntry(string filePath, IEnumerable<CompilationDatabaseEntry> compilationDatabaseEntries)
+        {
+            logger.LogDebug($"[CompilationConfigProvider] Code file detected, searching for exact match. File: {filePath}");
+
+            var entry = LocateCodeEntry(filePath, compilationDatabaseEntries);
+
+            if (entry == null)
+            {
+                logger.WriteLine(Resources.NoCompilationDatabaseEntry, filePath);
+            }
+
+            return entry;
+        }
+
+        private CompilationDatabaseEntry LocateMatchingCodeEntry(string headerFilePath, IEnumerable<CompilationDatabaseEntry> compilationDatabaseEntries)
```

Comment:
I can't think of a nicer way to structure this block 🙈

Reply:
One option would be to push more logic into each "Locate" method, so they handle the loop and the "not found" logging, returning null if they didn't find anything.

That would simplify the call to something like:
```
  var match = LocateCodeEntry(codeFile, compilationDatabaseEntries) ??
    LocateFirstCodeEntryWithSameName(codeFile, compilationDatabaseEntries) ??
    LocateFirstCodeEntryUnderRoot(codeFile, compilationDatabaseEntries) ??
    LocateFirstCodeEntry(compilationDatabaseEntries);
```

You'd probably also want to pass `codeFilesWithSameNameAndSamePath` into the "Locate" methods so they didn't have to recalculate it each time.

I'm not sure it's a big improvement, though.